### ISSUE TITLE
web ui: status?

### ISSUE THREAD

**Issue by mhaberler** [ElektraInitiative/libelektra#252]
is there such a thing as a web UI to browse/edit a store?

**Comment by markus2330** [comment]
No (there was something in python some time ago), but the new gui is qml, so it should be possible to port it. Any plans on that?

**Comment by mhaberler** [comment]
Not in scope for us - we would be primarily users as we are looking at filling holes by importing a package which fits; bugfixes are ok but I do not see our resources to fill in basic capabilities in libelektra

**Comment by markus2330** [comment]
A generic web-ui clearly is within Elektra's scope and could be implemented soon. But projects often want customized pages integrated in their already-existing web UI. What is the situation for your project?

**Comment by mhaberler** [comment]
What I would expect at this time a basic web service is available either through an internal mini-webserver along the lines of  node.js/libwebsockets/mongoose, or some integratable backend script which provides a websockets/JSON API to be used with a basic js client, for example something mundane like jquery.js . Probably a node.js npm is the fastest route to get there.

Functionally I would think the qt-gui is a good start.

That boils down to a (optionally authenticated) URI provided by the library if so configured/started. Any pages customization beyond that is out of scope AFAICT.

**Comment by markus2330** [comment]
What about the priority of functionality?

A suggestion:

1. basic editing+undo+import/export/search as in qt-gui
1. REST interface (curl)
1. basic authentication (mapping to local users)
1. encryption (hopefully for free anyway)
1. input validation of values
1. mounting+display help/plugin infos as in qt-gui
1. listener notification
1. improved authentication (roles, visibility,...)

Everything but 1. is somewhat optional for some type of users.

**Comment by mhaberler** [comment]
can I adjust priorities a bit:

1. fix notification, then the rest - polling for updates: I'm sorry, this is 2015 ;)
2. integrate elektra api into node.js as  npm 
you can fake the notification while working in parallel, but an notify fd in the node backend is key
3. once that works, drape whatever web-side is deemed appropriate - that is all very nice but the backend integration is the #1 deliverable of the whole effort

**Comment by markus2330** [comment]
I was only talking about use-cases of web ui. To clarify, I will split
7. listener notification

to:

a. reactive UI, immediately update keys when they were changed by another process
b. Comet/HTTP Push, immediately inform a REST user on KDB changes

b is somewhat an alternative to ZeroMQ. So I thought it would be less important for you to have it.

**Comment by omnidan** [comment]
If we use [redux](https://rackt.github.io/redux/) and [react](https://facebook.github.io/react/) together (with [react-redux](https://github.com/rackt/react-redux)) we get features like undo/import/export for free as we can [time-travel on the immutable states](https://github.com/gaearon/redux-devtools#redux-devtools-diff-monitor), for exporting/importing we can simply serialize the state into JSON. In addition to all that, we get real time updates of all UI components for free (thanks to reactivity).

The simplest approach would be sending actions directly to the web ui (maybe via websockets) and then handling them via redux.

**Comment by markus2330** [comment]
http API discussions happen in #921 and #912

**Comment by markus2330** [comment]
web-ui is now tracked in https://github.com/ElektraInitiative/libelektra/projects/4
